### DISCUSSION TITLE
Entt and collision detection

### DISCUSSION THREAD

**Discussion by Adoro** [skypjack/entt#677] [Q&A] [answered]
I'm pretty new to ECS and just learned about entt and want to try to implement it in some game engine code I wrote. I tried to google on how to deal with collision detection in an ECS but couldn't really find a good article about it. A lot of solutions seem to iterate over all entities and just check all bounds (talking about 2D here), which seems really stupid to me. 

What I'm thinking I should deal with is:

- Only check objects which have moved
- Only check objects which are in close proximity of the moved object, dont need to check objects on the other side of the screen.

So my question is, how can you use entt to only get objects in close proximity? If possible?
For the objects which have moved I reckon you could add some kind of listener / observer which observes the Transform component?

Pretty much my question is, how would you generally work with collision detection in entt? My question is mostly about 2D now but I think the solution would work for both 2D and 3D all the same.

**Comment by skypjack** [comment, marked as answer]
Using a component based design doesn't mean that you've to do **everything** in terms of entities and components. In this case you want to perform a spatial lookup and a linear array in which things are moved to optimize linear accesses isn't the best choice. EnTT allows to create custom pools with which you can structure your data to easy the job, but that's an implementation detail and something pretty advanced.
In general, I'd suggest to still use something like a grid based lookup, a quad tree or any other approach suitable for a broad phase and the algorithm that fits best with your requirements (AABB or whatever) for the narrow phase. You can put entities in these data structures and use them to get your data from the registry or use the data in the registry to for example update the leaf of an entity when it moves around.
Static vs moving elements is another optimization that makes sense but has nothing to do with the registry itself.

**Reply by hugoc7** [reply]
Is it a good idea to sort the components arrays according to space partitions ? It would avoid indirections when testing the collisions with all the entites inside a space partition, but is it worth it ? For example, is it used in Minecraft ? (maybe you're not supposed to say that)

**Reply by skypjack** [reply]
I wouldn't suggest it to be honest. Sorting your pools every tick is (very) likely more expensive than keeping an external ds up-to-date.
If you want to create partitions as pools into the registry, EnTT makes you create runtime pools. It may work for sure but how better is it if compared to a spatial index? Also, how optimized is it at the end of the day?
Every problem has its solution. Be wary of those who suggest hammering everything into some components and entities. It's not necessarily the right solution in every case.

**Comment by Green-Sky** [comment]
I hope @skypjack 's answer was to your satisfaction. I still would like to give you some pointers...:
As @skypjack said, the Registry is not a "one size fits all" solution, but it is a very solid base. Fast physics relies on a good "acceleration-structure", usually a "BVH", [(Bounding-Volume-Hierarchy)](https://en.wikipedia.org/wiki/Bounding_volume_hierarchy). @skypjack called this a "grid based lookup". I suggest you check out [Box2D](https://github.com/erincatto/box2d) if you want a ready implementaion, or just to look at a good implementation. If you are more into homebrew, you can also look at/start with [Box2D-lite](https://github.com/erincatto/box2d-lite) as a base. It is the result of a GDC workshop by the same author as Box2D. He also has other good GDC talks about physics.
There also exists [Edyn](https://github.com/xissburg/edyn), which is a wip physics implementation on top off EnTT.

edit: [Dynamic Bounding Volume Hierarchies — GDC 2019](https://box2d.org/files/ErinCatto_DynamicBVH_Full.pdf)

**Comment by Adoro** [comment]
Thank you for your responses, it helped me to give some perspective on how to implement this. I've been looking into Box2D and will be implementing that together with entt.
